File: src/engine/sanitize.ts
// Lore arrives from strangers. It is cleaned before display but the token is
// always kept in the dataset — dropping flagged rows would bias the sample.

// Zero-width spaces/joiners, BOM, and bidi controls (built from code points so the source stays ASCII).
const INVISIBLE_RANGES: [number, number][] = [
  [0x200b, 0x200f],
  [0x202a, 0x202e],
  [0x2060, 0x2069],
  [0xfeff, 0xfeff],
];
const ZERO_WIDTH_AND_BIDI = new RegExp(
  `[${INVISIBLE_RANGES.map(([a, b]) => `${String.fromCodePoint(a)}-${String.fromCodePoint(b)}`).join('')}]`,
  'g',
);
const URLS = /https?:\/\/\S+|ipfs:\/\/\S+|www\.\S+/gi;

/** Extend this list for production; it only needs to catch what you refuse to display. */
export const BLOCKLIST = new Set(['fuck', 'shit', 'bitch']);

export interface SanitizedLore {
  display: string;
  withheld: boolean;
  reason: 'blocklist' | null;
}

export function sanitizeLore(raw: string | null | undefined): SanitizedLore {
  if (!raw || !raw.trim()) return { display: '', withheld: false, reason: null };
  const text = raw.replace(ZERO_WIDTH_AND_BIDI, '');
  const words = text.toLowerCase().match(/\w+/g) ?? [];
  if (words.some((w) => BLOCKLIST.has(w))) return { display: '', withheld: true, reason: 'blocklist' };
  return { display: text.replace(URLS, '').replace(/\s{2,}/g, ' ').trim().slice(0, 280).trim(), withheld: false, reason: null };
}

File: backend/wassily/engine/sanitize.py
"""Lore arrives from strangers.

It is cleaned before display, but the token is always kept in the dataset:
dropping flagged rows would bias the sample.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

# Zero-width spaces/joiners, BOM, and bidi controls (escaped so the source stays ASCII).
_INVISIBLE = re.compile("[\\u200b-\\u200f\\u202a-\\u202e\\u2060-\\u2069\\ufeff]")
_URLS = re.compile(r"https?://\S+|ipfs://\S+|www\.\S+", re.IGNORECASE)
_WORDS = re.compile(r"\w+", re.ASCII)
_SPACES = re.compile(r"\s{2,}")

# Extend this list for production; it only needs to catch what you refuse to display.
BLOCKLIST = frozenset({"fuck", "shit", "bitch"})
MAX_LORE_CHARS = 280


@dataclass(frozen=True)
class SanitizedLore:
    display: str
    withheld: bool
    reason: Literal["blocklist"] | None


def sanitize_lore(raw: str | None) -> SanitizedLore:
    if not raw or not raw.strip():
        return SanitizedLore("", False, None)
    text = _INVISIBLE.sub("", raw)
    if any(w in BLOCKLIST for w in _WORDS.findall(text.lower())):
        return SanitizedLore("", True, "blocklist")
    display = _SPACES.sub(" ", _URLS.sub("", text)).strip()[:MAX_LORE_CHARS].strip()
    return SanitizedLore(display, False, None)
